### ISSUE TITLE
Does CONNECT request mean WebTransport should use fetch?

### ISSUE THREAD

**Issue by annevk** [w3c/webtransport#272]
WebSocket uses fetch for the handshake, meaning all kinds of security checks automatically apply. Now that WebTransport has a similar handshake, perhaps it should use fetch to execute it?

See https://fetch.spec.whatwg.org/#concept-websocket-establish.

**Comment by ricea** [comment]
Agreed in principle. The details may differ. As with WebSocket, we'd like to avoid handshakes. Unlike WebSocket, we can probably avoid them even for private network access, since a WebTransport connection can't be used to forge an HTTP request. WebTransport throttling probably will be different from WebSocket, but it hasn't been designed yet.

**Comment by ricea** [comment]
I originally said "handshakes" instead of "preflights" in the above message. Corrected.

**Comment by letitz** [comment]
> a WebTransport connection can't be used to forge an HTTP request

Can an application build its own HTTP request, serialize it and send it over WebTransport? My ignorance as to the transport part of WebTransport is showing - maybe messages sent over WebTransport cannot be confused with HTTP/3 requests. I guess my question is: can you explain why that is true?

**Comment by annevk** [comment]
I think apart from the handshake which is ordinary HTTP (though also requires an HTTP/3 connection) it will use its own frame type (WEBTRANSPORT_STREAM, https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3).

**Comment by letitz** [comment]
Makes sense! In order to exempt WebTransport handshakes from Private Network Access preflights, I would like to understand what malicious users could inject into the handshake. Are there any user-controlled headers or header values, for example?

**Comment by yutakahirano** [comment]
cc: @vasilvv 

Is it good to call https://fetch.spec.whatwg.org/#concept-fetch?

1. We need to keep the HTTP/3 stream alive as long as the WebTransport session is alive. How can we say so with it?
2. Calling https://fetch.spec.whatwg.org/#concept-fetch implies we attach headers such as accept-language, user-agent and so on.

Having https://fetch.spec.whatwg.org/#concept-fetch support WebTransport would be good if WebTransport needed cookie / authentication / redirect / caching / CORS preflight, but WebTransport doesn't.

We need to check CSP, but that can be done in https://w3c.github.io/webtransport/#webtransport-constructor.

For user-controlled headers, please see #263.

**Comment by annevk** [comment]
WebTransport does need port blocking, but I plan on moving that to "obtain a connection" so that would work.

**Comment by yutakahirano** [comment]
I think we don't need to resolve this by the initial release.

**Comment by annevk** [comment]
https://github.com/w3c/webtransport/pull/309 would also be handled by Fetch, although the current setup around that is somewhat shaky admittedly.

**Comment by yutakahirano** [comment]
Hmm... I don't think Fetch's termination would be good for #309 (or #127), as WebTransport's termination is more complex than fetch's termination. WebTransport is similar to WebSocket which has [its own logic on unloading](https://html.spec.whatwg.org/multipage/browsing-the-web.html#unloading-document-cleanup-steps).

**Comment by jan-ivar** [comment]
@yutakahirano I notice the spec references fetch algorithms now. Is this issue still relevant?

**Comment by yutakahirano** [comment]
Currently WebTransport uses the ["obtain a conneciton"](https://fetch.spec.whatwg.org/#concept-connection-obtain) procedure. This issue is whether we want to use the ["fetch"](https://fetch.spec.whatwg.org/#concept-fetch) procedure. So, they are different from each other.

**Comment by jan-ivar** [comment]
> Currently WebTransport uses the ["obtain a conneciton"](https://fetch.spec.whatwg.org/#concept-connection-obtain) procedure. This issue is whether we want to use the ["fetch"](https://fetch.spec.whatwg.org/#concept-fetch) procedure. So, they are different from each other.

@annevk it seems we interface with fetch at a low level here. Is this sufficient to provide the "kinds of security checks automatically" that you reference? If, not would you mind listing the missing ones? Thankls

**Comment by annevk** [comment]
I'm not sure which are relevant, but there is:

1. Referrer handling, including referrer policy enforcement
2. CSP
3. Port blocking (I hope to move this into connection establishment at which point it would be shared)
4. HSTS handling (hopefully not relevant as it's always secure?)
5. Mixed Content (hopefully not relevant as it's always secure?)

There's some other things there that's probably not relevant per @yutakahirano's comments above as the handshake does not follow redirects, does not have cookies, does not do HTTP auth, etc.

**Comment by jan-ivar** [comment]
@annevk Thanks, that seems to leave origin and csp.

Since we have separate issues on those two (https://github.com/w3c/webtransport/issues/59 on CSP) can we close this?

Our should we wait until you have moved port blocking?

**Comment by martinthomson** [comment]
There is no intent to send Referer with the request.  Origin should suffice (and we have a plan for fixing #368).

**Comment by jan-ivar** [comment]
We're already tracking
* https://github.com/w3c/webtransport/issues/59
* https://github.com/w3c/webtransport/issues/229
* https://github.com/w3c/webtransport/issues/368

@annevk is this sufficient to close this issue?
